File: apps/ios/Theora/Services/BrainClient.swift
import Foundation
import Combine

// MARK: - Connection State

enum BrainConnectionState: String, Equatable {
    case disconnected
    case connecting
    case registering
    case connected
    case reconnecting
}

// MARK: - Message Types

struct BrainTextResponse {
    let text: String
    let sdui: [String: Any]?
}

struct BrainTranscript {
    let text: String
    let isFinal: Bool
}

// MARK: - Brain Client (adapted from ASOSBrainClient)

final class BrainClient: NSObject, ObservableObject {

    @Published var connectionState: BrainConnectionState = .disconnected
    @Published var lastTextResponse: String = ""
    @Published var lastTranscript: BrainTranscript?
    @Published var streamDelta: String = ""

    var onTextResponse: ((BrainTextResponse) -> Void)?
    var onStreamDelta: ((String) -> Void)?
    var onAudioResponse: ((Data, String, Int, Bool) -> Void)?
    var onStopPlayback: (() -> Void)?

    private var host: String
    private var port: Int
    private var apiKey: String
    private let nodeId: String

    private var urlSession: URLSession?
    private var webSocket: URLSessionWebSocketTask?
    private var sessionId: String?

    private var reconnectAttempts = 0
    private let maxReconnectAttempts = 10
    private var sensorBuffer: [[String: Any]] = []
    private var flushTimer: Timer?
    private let sendQueue = DispatchQueue(label: "io.theora.brain.send", qos: .userInitiated)

    var isConnected: Bool { connectionState == .connected }

    init(host: String = "localhost", port: Int = 9090, apiKey: String = "") {
        self.host = host
        self.port = port
        self.apiKey = apiKey
        #if os(iOS)
        self.nodeId = "theora-iphone-\(UIDevice.current.name.lowercased().replacingOccurrences(of: " ", with: "-"))"
        #else
        self.nodeId = "theora-apple-device"
        #endif
        super.init()
    }

    // MARK: - Public API

    func configure(host: String, port: Int, apiKey: String) {
        self.host = host
        self.port = port
        self.apiKey = apiKey
    }

    func connect() {
        guard connectionState != .connected && connectionState != .connecting else { return }
        connectionState = .connecting

        let config = URLSessionConfiguration.default
        config.waitsForConnectivity = true
        urlSession = URLSession(configuration: config, delegate: self, delegateQueue: nil)

        let urlString = "ws://\(host):\(port)/v1/node?api_key=\(apiKey)"
        guard let url = URL(string: urlString) else {
            connectionState = .disconnected
            return
        }

        var request = URLRequest(url: url)
        request.timeoutInterval = 15
        webSocket = urlSession?.webSocketTask(with: request)
        webSocket?.resume()
        startReceiving()
    }

    func disconnect() {
        flushTimer?.invalidate()
        webSocket?.cancel(with: .normalClosure, reason: nil)
        DispatchQueue.main.async { self.connectionState = .disconnected }
    }

    func sendTextCommand(_ text: String) {
        let msg: [String: Any] = [
            "hop": "node",
            "type": "text_command",
            "payload": [
                "text": text,
                "node_id": nodeId
            ]
        ]
        sendJSON(msg)
    }

    func sendAudioChunk(base64: String, chunkIndex: Int, isFinal: Bool = false) {
        let msg: [String: Any] = [
            "hop": "node",
            "type": "audio_chunk",
            "payload": [
                "node_id": nodeId,
                "data_b64": base64,
                "chunk_index": chunkIndex,
                "is_final": isFinal,
                "encoding": "pcm16",
                "sample_rate": 24000
            ]
        ]
        sendJSON(msg)
    }

    func sendSensorData(sensor: String, value: [String: Any]) {
        let telemetry: [String: Any] = [
            "hop": "node",
            "type": "sensor_telemetry",
            "payload": [
                "node_id": nodeId,
                "sensor": sensor,
                "data": value,
                "timestamp": ISO8601DateFormatter().string(from: Date()),
                "source": "healthkit"
            ]
        ]
        sensorBuffer.append(telemetry)
    }

    func sendBatchSensorData(_ readings: [String: Any]) {
        let batch: [String: Any] = [
            "hop": "node",
            "type": "sensor_batch",
            "payload": [
                "node_id": nodeId,
                "readings": readings,
                "timestamp": ISO8601DateFormatter().string(from: Date())
            ]
        ]
        sendJSON(batch)
    }

    // MARK: - Registration

    private func registerAsNode() {
        connectionState = .registering

        var capabilities = ["microphone", "gps", "accelerometer", "gyroscope"]
        #if os(iOS)
        capabilities.append(contentsOf: ["camera", "healthkit"])
        #endif

        let registration: [String: Any] = [
            "hop": "node",
            "type": "register",
            "payload": [
                "node_id": nodeId,
                "node_type": "phone",
                "capabilities": capabilities,
                "platform": "ios",
                "os_version": ProcessInfo.processInfo.operatingSystemVersionString,
                "glasses_connected": false
            ]
        ]
        sendJSON(registration)
    }

    private func sendVoiceConfig() {
        let config: [String: Any] = [
            "hop": "node",
            "type": "voice_config",
            "payload": [
                "node_id": nodeId,
                "supports_realtime": true,
                "mode": "realtime",
                "sample_rate": 24000,
                "encoding": "pcm16"
            ]
        ]
        sendJSON(config)
    }

    // MARK: - WebSocket receive loop

    private func startReceiving() {
        webSocket?.receive { [weak self] result in
            guard let self = self else { return }
            switch result {
            case .success(let message):
                switch message {
                case .string(let text):
                    self.handleMessage(text)
                case .data(let data):
                    if let text = String(data: data, encoding: .utf8) {
                        self.handleMessage(text)
                    }
                @unknown default:
                    break
                }
                self.startReceiving()

            case .failure(let error):
                self.handleDisconnect(reason: error.localizedDescription)
            }
        }
    }

    private func handleMessage(_ text: String) {
        guard let data = text.data(using: .utf8),
              let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else { return }

        let type = json["type"] as? String ?? ""
        let payload = json["payload"] as? [String: Any] ?? [:]

        DispatchQueue.main.async { [weak self] in
            guard let self = self else { return }

            switch type {
            case "registered":
                self.connectionState = .connected
                self.sessionId = json["session_id"] as? String ?? ""
                self.reconnectAttempts = 0
                self.startSensorFlushTimer()
                self.sendVoiceConfig()

            case "text_response":
                let responseText = payload["text"] as? String ?? ""
                let sdui = payload["sdui"] as? [String: Any]
                self.lastTextResponse = responseText
                self.onTextResponse?(BrainTextResponse(text: responseText, sdui: sdui))

            case "stream_delta":
                let delta = payload["delta"] as? String ?? payload["text"] as? String ?? ""
                self.streamDelta += delta
                self.onStreamDelta?(delta)

            case "stream_end":
                let finalText = self.streamDelta
                self.lastTextResponse = finalText
                self.onTextResponse?(BrainTextResponse(text: finalText, sdui: nil))
                self.streamDelta = ""

            case "audio_response":
                let audioB64 = payload["data_b64"] as? String ?? ""
                let encoding = payload["encoding"] as? String ?? "pcm16"
                let sampleRate = payload["sample_rate"] as? Int ?? 24000
                let isFinal = payload["is_final"] as? Bool ?? false
                if let audioData = Data(base64Encoded: audioB64) {
                    self.onAudioResponse?(audioData, encoding, sampleRate, isFinal)
                }

            case "speech_started":
                self.onStopPlayback?()

            case "transcript":
                let transcriptText = payload["text"] as? String ?? ""
                let isPartial = payload["is_partial"] as? Bool ?? false
                self.lastTranscript = BrainTranscript(text: transcriptText, isFinal: !isPartial)

            default:
                break
            }
        }
    }

    // MARK: - Reconnection

    private func handleDisconnect(reason: String) {
        flushTimer?.invalidate()
        DispatchQueue.main.async { self.connectionState = .disconnected }
        attemptReconnect()
    }

    private func attemptReconnect() {
        guard reconnectAttempts < maxReconnectAttempts else { return }
        DispatchQueue.main.async { self.connectionState = .reconnecting }
        reconnectAttempts += 1

        let delay = min(Double(reconnectAttempts) * 2.0, 30.0)
        DispatchQueue.main.asyncAfter(deadline: .now() + delay) { [weak self] in
            guard let self = self, self.connectionState == .reconnecting else { return }
            self.connect()
        }
    }

    // MARK: - Send helpers

    private func sendJSON(_ dict: [String: Any]) {
        sendQueue.async { [weak self] in
            guard let data = try? JSONSerialization.data(withJSONObject: dict),
                  let text = String(data: data, encoding: .utf8) else { return }
            self?.webSocket?.send(.string(text)) { _ in }
        }
    }

    private func startSensorFlushTimer() {
        DispatchQueue.main.async {
            self.flushTimer?.invalidate()
            self.flushTimer = Timer.scheduledTimer(withTimeInterval: 2.0, repeats: true) { [weak self] _ in
                self?.flushSensorBuffer()
            }
        }
    }

    private func flushSensorBuffer() {
        guard !sensorBuffer.isEmpty else { return }
        let batch = sensorBuffer
        sensorBuffer.removeAll()
        for msg in batch { sendJSON(msg) }
    }
}

// MARK: - URLSessionWebSocketDelegate

extension BrainClient: URLSessionWebSocketDelegate {
    func urlSession(_ session: URLSession, webSocketTask: URLSessionWebSocketTask,
                    didOpenWithProtocol protocol: String?) {
        registerAsNode()
    }

    func urlSession(_ session: URLSession, webSocketTask: URLSessionWebSocketTask,
                    didCloseWith closeCode: URLSessionWebSocketTask.CloseCode, reason: Data?) {
        let reasonStr = reason.flatMap { String(data: $0, encoding: .utf8) } ?? "Connection closed"
        handleDisconnect(reason: reasonStr)
    }
}

File: apps/android/app/src/main/java/io/feral/app/services/BrainClient.kt
package ai.feral.app.services

import kotlinx.coroutines.*
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.serialization.json.*
import okhttp3.*
import java.util.concurrent.TimeUnit

enum class BrainConnectionState {
    DISCONNECTED, CONNECTING, REGISTERING, CONNECTED, RECONNECTING
}

data class ChatMessage(
    val text: String,
    val isUser: Boolean,
    val timestamp: Long = System.currentTimeMillis(),
)

class BrainClient(
    private var host: String = "10.0.2.2",
    private var port: Int = 9090,
    private var apiKey: String = "",
    private val nodeId: String = "feral-android-${android.os.Build.MODEL.lowercase().replace(" ", "-")}",
) {
    private var webSocket: WebSocket? = null
    private val httpClient = OkHttpClient.Builder()
        .readTimeout(0, TimeUnit.MILLISECONDS)
        .build()
    private val scope = CoroutineScope(Dispatchers.IO + SupervisorJob())
    private val json = Json { ignoreUnknownKeys = true; isLenient = true }

    private val _connectionState = MutableStateFlow(BrainConnectionState.DISCONNECTED)
    val connectionState: StateFlow<BrainConnectionState> = _connectionState

    private val _messages = MutableStateFlow<List<ChatMessage>>(emptyList())
    val messages: StateFlow<List<ChatMessage>> = _messages

    private val _streamingText = MutableStateFlow("")
    val streamingText: StateFlow<String> = _streamingText

    private val _transcript = MutableStateFlow("")
    val transcript: StateFlow<String> = _transcript

    private var reconnectAttempts = 0
    private val maxReconnectAttempts = 10

    val isConnected: Boolean get() = _connectionState.value == BrainConnectionState.CONNECTED

    fun configure(host: String, port: Int, apiKey: String) {
        this.host = host
        this.port = port
        this.apiKey = apiKey
    }

    fun connect() {
        if (_connectionState.value == BrainConnectionState.CONNECTED ||
            _connectionState.value == BrainConnectionState.CONNECTING) return

        _connectionState.value = BrainConnectionState.CONNECTING
        val url = "ws://$host:$port/v1/node?api_key=$apiKey"
        val request = Request.Builder().url(url).build()

        webSocket = httpClient.newWebSocket(request, object : WebSocketListener() {
            override fun onOpen(webSocket: WebSocket, response: Response) {
                _connectionState.value = BrainConnectionState.REGISTERING
                sendRegistration()
                sendVoiceConfig()
                _connectionState.value = BrainConnectionState.CONNECTED
                reconnectAttempts = 0
            }

            override fun onMessage(webSocket: WebSocket, text: String) {
                scope.launch { handleMessage(text) }
            }

            override fun onClosing(webSocket: WebSocket, code: Int, reason: String) {
                _connectionState.value = BrainConnectionState.DISCONNECTED
            }

            override fun onFailure(webSocket: WebSocket, t: Throwable, response: Response?) {
                _connectionState.value = BrainConnectionState.DISCONNECTED
                scheduleReconnect()
            }
        })
    }

    fun disconnect() {
        webSocket?.close(1000, "Client disconnect")
        _connectionState.value = BrainConnectionState.DISCONNECTED
    }

    fun sendTextCommand(text: String) {
        _messages.value = _messages.value + ChatMessage(text = text, isUser = true)
        _streamingText.value = ""

        val msg = buildJsonObject {
            put("type", "text_command")
            put("hop", "node")
            putJsonObject("payload") {
                put("text", text)
                put("node_id", nodeId)
            }
        }
        sendJson(msg)
    }

    fun sendAudioChunk(data: ByteArray, chunkIndex: Int, isFinal: Boolean) {
        val b64 = android.util.Base64.encodeToString(data, android.util.Base64.NO_WRAP)
        val msg = buildJsonObject {
            put("type", "audio_chunk")
            put("hop", "node")
            putJsonObject("payload") {
                put("node_id", nodeId)
                put("data_b64", b64)
                put("chunk_index", chunkIndex)
                put("is_final", isFinal)
                put("encoding", "pcm16")
                put("sample_rate", 24000)
            }
        }
        sendJson(msg)
    }

    fun sendSensorTelemetry(
        heartRate: Int? = null,
        spo2: Int? = null,
        steps: Int? = null,
        temperature: Double? = null,
    ) {
        val msg = buildJsonObject {
            put("type", "sensor_telemetry")
            put("hop", "node")
            putJsonObject("payload") {
                put("node_id", nodeId)
                heartRate?.let { put("heart_rate", it) }
                spo2?.let { put("spo2_pct", it) }
                steps?.let { put("step_count", it) }
                temperature?.let { put("skin_temp_c", it) }
            }
        }
        sendJson(msg)
    }

    private fun sendRegistration() {
        val msg = buildJsonObject {
            put("type", "register")
            put("hop", "node")
            putJsonObject("payload") {
                put("node_id", nodeId)
                put("node_type", "phone")
                put("platform", "android")
                put("os_version", android.os.Build.VERSION.RELEASE)
                put("device_model", android.os.Build.MODEL)
                putJsonArray("capabilities") {
                    add("voice"); add("sensors"); add("gps"); add("camera"); add("health_connect")
                }
            }
        }
        sendJson(msg)
    }

    private fun sendVoiceConfig() {
        val msg = buildJsonObject {
            put("type", "voice_config")
            put("hop", "node")
            putJsonObject("payload") {
                put("node_id", nodeId)
                put("supports_realtime", true)
                put("mode", "realtime")
                put("sample_rate", 24000)
                put("encoding", "pcm16")
            }
        }
        sendJson(msg)
    }

    private fun handleMessage(text: String) {
        try {
            val obj = json.parseToJsonElement(text).jsonObject
            val type = obj["type"]?.jsonPrimitive?.contentOrNull ?: return
            val payload = obj["payload"]?.jsonObject ?: JsonObject(emptyMap())

            when (type) {
                "registered" -> {
                    _connectionState.value = BrainConnectionState.CONNECTED
                }

                "text_response" -> {
                    val msg = payload["text"]?.jsonPrimitive?.contentOrNull ?: ""
                    _messages.value = _messages.value + ChatMessage(text = msg, isUser = false)
                    _streamingText.value = ""
                }

                "stream_delta" -> {
                    val delta = payload["delta"]?.jsonPrimitive?.contentOrNull
                        ?: payload["text"]?.jsonPrimitive?.contentOrNull ?: ""
                    _streamingText.value += delta
                }

                "stream_end" -> {
                    val final = _streamingText.value
                    if (final.isNotEmpty()) {
                        _messages.value = _messages.value + ChatMessage(text = final, isUser = false)
                    }
                    _streamingText.value = ""
                }

                "transcript" -> {
                    val t = payload["text"]?.jsonPrimitive?.contentOrNull ?: ""
                    _transcript.value = t
                }

                "audio_response", "speech_started", "execute" -> {
                    // Audio playback and execution handling delegated to platform layer
                }
            }
        } catch (_: Exception) {}
    }

    private fun sendJson(json: JsonObject) {
        webSocket?.send(json.toString())
    }

    private fun scheduleReconnect() {
        if (reconnectAttempts >= maxReconnectAttempts) return
        _connectionState.value = BrainConnectionState.RECONNECTING
        reconnectAttempts++
        val delay = minOf(reconnectAttempts * 2000L, 30_000L)
        scope.launch {
            delay(delay)
            if (_connectionState.value == BrainConnectionState.RECONNECTING) {
                connect()
            }
        }
    }

    fun clearMessages() {
        _messages.value = emptyList()
    }

    fun destroy() {
        disconnect()
        scope.cancel()
    }
}
